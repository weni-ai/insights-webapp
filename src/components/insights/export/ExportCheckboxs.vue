<template>
  <section class="form-checkboxs-data">
    <section class="form-checkboxs-data__models">
      <template
        v-for="(modelData, modelName) in modelFields"
        :key="modelName"
      >
        <section
          v-if="shouldRenderModel(String(modelName))"
          class="form-checkboxs-data__model"
        >
          <div class="form-checkboxs-data__model-header">
            <UnnnicCheckbox
              :modelValue="isModelEnabled(String(modelName))"
              :textRight="getModelLabel(String(modelName))"
              @change="handleModelToggle(String(modelName), $event)"
            />
          </div>

          <div
            v-if="
              isModelEnabled(String(modelName)) &&
              getModelFieldsList(String(modelName)).length > 0
            "
            class="form-checkboxs-data__fields"
          >
            <template
              v-for="field in getModelFieldsList(String(modelName))"
              :key="field.name"
            >
              <UnnnicCheckbox
                :modelValue="isFieldSelected(String(modelName), field.name)"
                :textRight="getFieldLabel(field.name)"
                @change="
                  handleFieldToggle(String(modelName), field.name, $event)
                "
              />
            </template>
          </div>
        </section>
      </template>
    </section>
    <section
      v-if="isLoading"
      class="form-checkboxs-data__loading"
    >
      <UnnnicSkeletonLoading
        v-for="i in 5"
        :key="i"
        width="100%"
        height="26px"
        data-testid="form-checkboxs-data-loading"
      />
    </section>
  </section>
</template>

<script setup lang="ts">
import { useI18n } from 'vue-i18n';

// Types
interface ModelField {
  type?: string;
}

interface ModelFields {
  [modelName: string]: {
    [fieldName: string]: ModelField;
  };
}

interface FieldItem {
  name: string;
  type?: string;
}

interface FilterItem {
  value: string;
  label: string;
}

interface ModelFilterConfig {
  modelName: string;
  filterData: FilterItem[];
}

// Props
interface Props {
  modelFields: ModelFields;
  selectedFields: { [modelName: string]: string[] };
  enabledModels: string[];
  modelFilters: ModelFilterConfig[];
  isLoading: boolean;
}

const props = withDefaults(defineProps<Props>(), {
  modelFields: () => ({}),
  selectedFields: () => ({}),
  enabledModels: () => [],
  modelFilters: () => [],
  isLoading: false,
});

// Emits
const emit = defineEmits<{
  'model-toggle': [modelName: string, enabled: boolean];
  'field-toggle': [modelName: string, fieldName: string, selected: boolean];
}>();

const { t } = useI18n();

const getModelFieldsList = (modelName: string): FieldItem[] => {
  const modelData = props.modelFields[modelName];
  if (!modelData) return [];

  return Object.entries(modelData).map(([fieldName, fieldData]) => ({
    name: fieldName,
    type: fieldData.type || 'string',
  }));
};

const isFieldSelected = (modelName: string, fieldName: string): boolean => {
  return props.selectedFields[modelName]?.includes(fieldName) || false;
};

const isModelEnabled = (modelName: string): boolean => {
  return props.enabledModels.includes(modelName);
};

const shouldRenderModel = (modelName: string): boolean => {
  const filterConfig = props.modelFilters.find(
    (filter) => filter.modelName === modelName,
  );

  if (filterConfig) {
    return filterConfig.filterData.length > 0;
  }

  return true;
};

const handleModelToggle = (modelName: string, enabled: boolean): void => {
  emit('model-toggle', modelName, enabled);
};

const handleFieldToggle = (
  modelName: string,
  fieldName: string,
  selected: boolean,
): void => {
  emit('field-toggle', modelName, fieldName, selected);
};

const getModelLabel = (modelName: string): string => {
  const translationKey = `export_data.model_labels.${modelName}`;
  const translation = t(translationKey);

  if (translation && translation !== translationKey) {
    return translation;
  }

  return modelName.replace(/_/g, ' ').replace(/\b\w/g, (l) => l.toUpperCase());
};

const getFieldLabel = (fieldName: string): string => {
  const translationKey = `export_data.field_labels.${fieldName}`;
  const translation = t(translationKey);

  let fieldLabel;

  if (translation && translation !== translationKey) {
    fieldLabel = translation;
  } else {
    fieldLabel = fieldName
      .replace(/_/g, ' ')
      .replace(/\b\w/g, (l) => l.toUpperCase());
  }

  return fieldLabel;
};
</script>

<style scoped lang="scss">
.form-checkboxs-data {
  display: flex;
  flex-direction: column;
  gap: $unnnic-spacing-sm;

  &__models {
    display: flex;
    flex-direction: column;
    gap: $unnnic-spacing-md;
  }

  &__model {
    &-header {
      margin-bottom: $unnnic-spacing-xs;

      :deep(.unnnic-checkbox) {
        font-weight: $unnnic-font-weight-bold;
        color: $unnnic-color-neutral-darkest;
      }
    }
  }

  &__fields {
    display: flex;
    flex-direction: column;
    gap: $unnnic-spacing-xs;
    padding-left: $unnnic-spacing-md;
    margin-left: $unnnic-spacing-xs;

    :deep(.unnnic-checkbox) {
      font-size: $unnnic-font-size-body-md;
      color: $unnnic-color-neutral-dark;
    }
  }

  &__loading {
    display: flex;
    flex-direction: column;
    gap: $unnnic-spacing-sm;
  }
}
</style>
