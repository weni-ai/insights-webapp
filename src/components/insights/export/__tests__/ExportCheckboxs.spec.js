import { beforeEach, describe, expect, it } from 'vitest';
import { config, mount } from '@vue/test-utils';
import { createI18n } from 'vue-i18n';

import ExportCheckboxs from '../ExportCheckboxs.vue';

const i18n = createI18n({
  legacy: false,
  locale: 'en',
  messages: {
    en: {
      export_data: {
        model_labels: {
          user: 'User',
          contact: 'Contact',
        },
        field_labels: {
          email: 'Email',
          name: 'Name',
          phone: 'Phone',
        },
      },
    },
  },
  fallbackWarn: false,
  missingWarn: false,
});

config.global.plugins = [i18n];

describe('ExportCheckboxs', () => {
  let wrapper;

  const mockModelFields = {
    user: {
      email: { type: 'string' },
      name: { type: 'string' },
    },
    contact: {
      phone: { type: 'string' },
      name: { type: 'string' },
    },
  };

  const defaultProps = {
    modelFields: mockModelFields,
    selectedFields: { user: ['email'] },
    enabledModels: ['user'],
    modelFilters: [],
    isLoading: false,
  };

  const createWrapper = (props = {}) => {
    return mount(ExportCheckboxs, {
      props: { ...defaultProps, ...props },
      global: {
        stubs: {
          UnnnicCheckbox: true,
          UnnnicSkeletonLoading: true,
        },
      },
    });
  };

  const section = () => wrapper.find('.form-checkboxs-data');

  beforeEach(() => {
    wrapper = createWrapper();
  });

  describe('Component rendering', () => {
    it('should render main section', () => {
      expect(section().exists()).toBe(true);
    });

    it('should render models section', () => {
      const modelsSection = wrapper.find('.form-checkboxs-data__models');
      expect(modelsSection.exists()).toBe(true);
    });

    it('should render loading skeletons when isLoading is true', () => {
      wrapper = createWrapper({ isLoading: true });
      const loadingSection = wrapper.find('.form-checkboxs-data__loading');
      expect(loadingSection.exists()).toBe(true);
    });

    it('should not render loading when isLoading is false', () => {
      const loadingSection = wrapper.find('.form-checkboxs-data__loading');
      expect(loadingSection.exists()).toBe(false);
    });
  });

  describe('Props', () => {
    it('should use default values', () => {
      wrapper = createWrapper({
        modelFields: undefined,
        selectedFields: undefined,
        enabledModels: undefined,
        modelFilters: undefined,
        isLoading: undefined,
      });

      expect(wrapper.props('modelFields')).toEqual({});
      expect(wrapper.props('selectedFields')).toEqual({});
      expect(wrapper.props('enabledModels')).toEqual([]);
      expect(wrapper.props('modelFilters')).toEqual([]);
      expect(wrapper.props('isLoading')).toBe(false);
    });

    it('should accept complex model fields structure', () => {
      expect(wrapper.props('modelFields')).toEqual(mockModelFields);
    });
  });

  describe('Model helpers', () => {
    it('should check if model is enabled', () => {
      expect(wrapper.vm.isModelEnabled('user')).toBe(true);
      expect(wrapper.vm.isModelEnabled('contact')).toBe(false);
    });

    it('should get model fields list', () => {
      const fields = wrapper.vm.getModelFieldsList('user');
      expect(fields).toHaveLength(2);
      expect(fields[0]).toEqual({ name: 'email', type: 'string' });
    });

    it('should return empty array for non-existent model', () => {
      const fields = wrapper.vm.getModelFieldsList('nonexistent');
      expect(fields).toEqual([]);
    });

    it('should check if field is selected', () => {
      expect(wrapper.vm.isFieldSelected('user', 'email')).toBe(true);
      expect(wrapper.vm.isFieldSelected('user', 'name')).toBe(false);
    });

    it('should return false for non-existent model in field selection', () => {
      expect(wrapper.vm.isFieldSelected('nonexistent', 'field')).toBe(false);
    });
  });

  describe('Model filtering', () => {
    it('should render model when no filter is configured', () => {
      expect(wrapper.vm.shouldRenderModel('user')).toBe(true);
    });

    it('should render model when filter has data', () => {
      wrapper = createWrapper({
        modelFilters: [
          {
            modelName: 'user',
            filterData: [{ value: '1', label: 'Option 1' }],
          },
        ],
      });
      expect(wrapper.vm.shouldRenderModel('user')).toBe(true);
    });

    it('should not render model when filter is empty', () => {
      wrapper = createWrapper({
        modelFilters: [{ modelName: 'user', filterData: [] }],
      });
      expect(wrapper.vm.shouldRenderModel('user')).toBe(false);
    });
  });

  describe('Label formatting', () => {
    it('should return translated model label', () => {
      expect(wrapper.vm.getModelLabel('user')).toBe('User');
    });

    it('should format model name when translation is missing', () => {
      expect(wrapper.vm.getModelLabel('user_profile')).toBe('User Profile');
    });

    it('should return translated field label', () => {
      expect(wrapper.vm.getFieldLabel('email')).toBe('Email');
    });

    it('should format field name when translation is missing', () => {
      expect(wrapper.vm.getFieldLabel('first_name')).toBe('First Name');
    });

    it('should capitalize first letter of each word', () => {
      expect(wrapper.vm.getFieldLabel('contact_phone_number')).toBe(
        'Contact Phone Number',
      );
    });
  });

  describe('Events', () => {
    it('should emit model-toggle event', async () => {
      await wrapper.vm.handleModelToggle('contact', true);

      expect(wrapper.emitted('model-toggle')).toBeTruthy();
      expect(wrapper.emitted('model-toggle')[0]).toEqual(['contact', true]);
    });

    it('should emit field-toggle event', async () => {
      await wrapper.vm.handleFieldToggle('user', 'name', true);

      expect(wrapper.emitted('field-toggle')).toBeTruthy();
      expect(wrapper.emitted('field-toggle')[0]).toEqual([
        'user',
        'name',
        true,
      ]);
    });

    it('should handle multiple toggle events', async () => {
      await wrapper.vm.handleModelToggle('user', false);
      await wrapper.vm.handleModelToggle('contact', true);

      expect(wrapper.emitted('model-toggle')).toHaveLength(2);
    });
  });

  describe('Conditional rendering', () => {
    it('should render fields when model is enabled', () => {
      wrapper = createWrapper({
        enabledModels: ['user'],
        selectedFields: { user: ['email'] },
      });

      const modelSection = wrapper.find('.form-checkboxs-data__model');
      expect(modelSection.exists()).toBe(true);
    });

    it('should handle empty model fields', () => {
      wrapper = createWrapper({
        modelFields: { empty_model: {} },
        enabledModels: ['empty_model'],
      });

      expect(wrapper.vm.getModelFieldsList('empty_model')).toEqual([]);
    });
  });

  describe('Component structure', () => {
    it('should have correct CSS classes', () => {
      expect(section().classes()).toContain('form-checkboxs-data');
    });

    it('should match snapshot', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
