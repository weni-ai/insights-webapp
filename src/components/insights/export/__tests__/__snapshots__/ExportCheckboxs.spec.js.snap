// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExportCheckboxs > Component structure > should match snapshot 1`] = `
<section
  class="form-checkboxs-data"
  data-testid="form-checkboxs-data"
  data-v-b513ad74=""
>
  <section
    class="form-checkboxs-data__models"
    data-testid="form-checkboxs-data-models"
    data-v-b513ad74=""
  >
    
    
    <section
      class="form-checkboxs-data__model"
      data-testid="form-checkboxs-model-user"
      data-v-b513ad74=""
    >
      <div
        class="form-checkboxs-data__model-header"
        data-v-b513ad74=""
      >
        <unnnic-checkbox-stub
          data-testid="form-checkboxs-model-checkbox-user"
          data-v-b513ad74=""
          modelvalue="true"
          textright="User"
        />
      </div>
      <div
        class="form-checkboxs-data__fields"
        data-testid="form-checkboxs-fields-user"
        data-v-b513ad74=""
      >
        
        <unnnic-checkbox-stub
          data-testid="form-checkboxs-field-user-email"
          data-v-b513ad74=""
          modelvalue="true"
          textright="Email"
        />
        <unnnic-checkbox-stub
          data-testid="form-checkboxs-field-user-name"
          data-v-b513ad74=""
          modelvalue="false"
          textright="Name"
        />
        
      </div>
    </section>
    
    
    <section
      class="form-checkboxs-data__model"
      data-testid="form-checkboxs-model-contact"
      data-v-b513ad74=""
    >
      <div
        class="form-checkboxs-data__model-header"
        data-v-b513ad74=""
      >
        <unnnic-checkbox-stub
          data-testid="form-checkboxs-model-checkbox-contact"
          data-v-b513ad74=""
          modelvalue="false"
          textright="Contact"
        />
      </div>
      <!--v-if-->
    </section>
    
    
  </section>
  <!--v-if-->
</section>
`;
